ipset/ipset.go

brianonn opened the pull request:
To fix the ipset bug

brianonn committed:
ipset/ipset.go

mheese commented on ipset/ipset.go:
```diff
@@ -239,11 +236,27 @@ func (s *IPSet) Destroy() error {
-func DestroyAll() error {
+// The prefix may be a prefix string or the constant ipset.AllSets
+// to specify all all sets should be destroyed
+func DestroyAll(prefix string) error {
```
This is an API breaking change. I assume we are the only ones using this specific repo?

brianonn replied:
I honestly don't know if we are the only consumers but I think we may be.  

I had it originally in a second function `DestroyAllByPrefix(prefix string)`  and didn't touch this function.  Perhaps that  is the better way to go.  It means changing our call then, but it does protect the API contract. 

The other option I thought about was to use a non-breaking trick I did before.. 

```
func DestroyAll(prefix ...string) {
    if len(prefix) == 0 {
        realPrefix = ""
    } else { 
        realPrefix = prefix[0]
    }
}
```

brianonn replied:
I fixed it with the variadic style

mheese reviewed (approved):
Looks good if you answer "yes" to my question in line :)

brianonn committed:
fixed DestroyAll function so as to not break the API

The DestroyAll() function now works with no parameters again, in case
there was another consumer of the API, we do not want to break the API
contract. So I have used a variadic trick to allow the DestroyAll()
function to accept 0 or 1 arguments (or many, but all arguments after
the first one is ignored.

I also fixed up the list() fuction in this commit since it was not
listing all names as I expected, I misunderstood how it operated before
now, so I added a new private function listAllSetNames() to get the full
list of all current ipset names and use that in the DestroyAll() loop.

I decided to do DestroyAll one by one in a loop because although ipset
can take a `ipset destroy` command that will destroy all, its not clear
what its behavior will be when it fails on one set of the many, i.e.
will it stop, and fail, or will it continue with the remaining sets?
Even if I determine one way to be true today, it maybe changed in a
future version of ipset.  I felt it was better to destroy each ipset
one-by-one in a loop and get the individual success or error status of
each one as well.

brianonn committed:
Command.Output() to CombinedOutput()
